pulsar-client-kafka-compat/src/main/java/org/apache/kafka/clients/consumer/PulsarKafkaConsumer.java

```diff
@@ -0,0 +1,482 @@
+     *
+     * @return The set of topics currently subscribed to
+     */
+    @Override
+    public Set<String> subscription() {
+        return consumers.keySet();
+    }
+
+    @Override
+    public void subscribe(Collection<String> topics) {
+        topics.forEach(topic -> {
+            try {
+                // Create individual subscription on each partition, that way we can keep using the
+                // acknowledgeCumulative()
+                PartitionedTopicMetadata partitionMetadata = ((PulsarClientImpl) client)
+                        .getPartitionedTopicMetadata(topic).get();
+
+                ConsumerConfiguration conf = new ConsumerConfiguration();
+                conf.setSubscriptionType(SubscriptionType.Failover);
+                conf.setMessageListener(this);
+                if (partitionMetadata.partitions > 1) {
+                    // Subscribe to each partition
+                    conf.setConsumerName(ConsumerName.generateRandomName());
+                    for (int i = 0; i < partitionMetadata.partitions; i++) {
+                        String partitionName = DestinationName.get(topic).getPartition(i).toString();
+                        org.apache.pulsar.client.api.Consumer consumer = client.subscribe(partitionName, groupId, conf);
+                        consumers.put(partitionName, consumer);
+                    }
+                } else {
+                    // Topic has a single partition
+                    org.apache.pulsar.client.api.Consumer consumer = client.subscribe(topic, groupId, conf);
+                    consumers.put(topic, consumer);
+                }
+            } catch (InterruptedException | ExecutionException | PulsarClientException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    @Override
+    public void subscribe(Collection<String> topics, ConsumerRebalanceListener callback) {
+        throw new UnsupportedOperationException("ConsumerRebalanceListener is not supported");
+    }
+
+    @Override
+    public void assign(Collection<TopicPartition> partitions) {
+        throw new UnsupportedOperationException("Cannot manually assign partitions");
+    }
+
+    @Override
+    public void subscribe(Pattern pattern, ConsumerRebalanceListener callback) {
+        throw new UnsupportedOperationException("Cannot subscribe with topic name pattern");
+    }
+
+    @Override
+    public void unsubscribe() {
+        consumers.values().forEach(c -> {
+            try {
+                c.unsubscribe();
+            } catch (PulsarClientException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public ConsumerRecords<K, V> poll(long timeoutMillis) {
+        try {
+            QueueItem item = receivedMessages.poll(timeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
this can be improved in future: you can pull all the data in the buffer, and only wait if there is no data in the buffer.

Reply:
Yes, that is the current behavior on the blocking queue.